### ISSUE TITLE
Add supports for generics

### ISSUE THREAD

**Issue by a-peyrard** [restx/restx#159]
Almost everything is in the title ;)

A task like this will require an impact analysis, and also an analysis of what we really expect with a support of generics. 
I will try in the next days to think about what is doable, and we will see together if it might be interesting or not.

**Comment by fcamblor** [comment]
Generics on which places ?
- On in/out endpoint parameters ? (we already have [this pull request](https://github.com/restx/restx/pull/118), and addressing this topic more deeply [on this issue](https://github.com/restx/restx/issues/146) I started to work on :-))
- On Restx components ?

**Comment by xhanin** [comment]
I think it's too wide to just say support for generics. We should clarify
where we don't support them yet and where we want to improve that. I think
the first focus is in restx factory, supporting generics when both
publishing components, querying components and in dependency injection.
Currently components are identified with a Name which is composed of a
Class and a String. Because generics aren't reified the Class object is not
enough to capture the generics type. And therefore we later can't query
components based on their generics type, and can't get dependency injection
support injection based on generics.

One thing we have to think about also is that we currently support multi
injection of components based on the generics type of Collection or similar
interfaces. We can't break that, so we Ned to be careful on how we support
generics in injection to state that it won't work with interfaces used for
multi injection.

**Comment by a-peyrard** [comment]
So everything were not in the title, wrote this to fast, and didn't realise how vague it was.

As Xavier explained, what I want to analyse is bringing generics in dependency injection. In order to publish some components under their parameterized types, and to be able to query those components using those parameterized types.

(I will change the title of this issue).

For example, one could provides components like this:
```
@Provides
public ThingBuilder<Foo> fooBuilder(...)
```
```
@Provides
public ThingBuilder<Bar> fooBuilder(...)
```

And ask for injection them like that:
```
@Provides
public AComponent aComponent(ThingBuilder<Foo> fooBuilder) {
```

Of course we have to keep the multi-injection process, so types used in that list (https://github.com/restx/restx/blob/5afdd71e1dd849fa10424cc325351fd0032b13c7/restx-factory/src/main/java/restx/factory/processor/FactoryAnnotationProcessor.java#L561) would not be allowed to be provided.

Continuing on examples and multi-injection, there would be two cases of multi-injection:
```
@Provides
public MultiComponentA multiA(List<ThingBuilder> builders) {
```
and
```
@Provides
public MultiComponentB multiB(List<ThingBuilder<Foo>> fooBuilders) {
```

Other things, parameterized types can use mutliples types arguments like:
 `Tranformation<Foo, Bar>` and they also can be multi-level: `Builder<Thing<Foo>>`.
 
So first I propose to only concentrate on one level of parametrized types (single and multiple). 

WDYT and sorry for this unclear issue, didn't realize I was rocking the boat ;)

**Comment by fcamblor** [comment]
I believe this is something hard to implement, but I would love to have this feature in restx :-)

**Comment by a-peyrard** [comment]
I scratched the surface of the subject this morning, read things on the net, and looked at other open-sources projects, in order to see how they deal with that kind of problem.

It seems that they all use the same technical, described by Neal Gafter in this post: http://gafter.blogspot.fr/2006/12/super-type-tokens.html?m=1

So currently `Name` is used like a primary key for components, and is composed, by a `Class` and a `String`. I think that the `Class` attribute has to be changed to a `TypeReference`.
For the name of this new class, `TypeReference`, Gafter introduces the name, jackson-core is using this name, guice is using `TypeLiteral`, spring is using `ParameterizedTypeReference`...
Any idea for the name ? I think I will start with `TypeReference`.

Of course if the `Name`is changed, current API must not be broken, building a `Name` with a `Class` and a `String` must remain doable.

**Comment by xhanin** [comment]
For the name, I think `TypeReference` is fine. But I think that can be used only to easily capture a parameterized type, and in the `Name` API we could use directly the `java.lang.reflect.Type` type instead of `java.lang.Class`. The advantage is that it's almost compatible with current implementation, since a `Class` is a `Type`.  Of course because `Type` by itself doesn't define much we will have to support only a limited number of implementation / subclasses (probably only `java.lang.Class` and `java.lang.reflect.ParameterizedType`) but I think it's closer to core java API and more extensible.

And note also that besides the TypeReference that can be used to capture the parameterized type, you can also use the `Types` utils class of restx:
https://github.com/restx/restx/blob/fabe16b110b615db2a4cb8ce8584761f352cb6f1/restx-common/src/main/java/restx/common/Types.java

WDYT?

**Comment by a-peyrard** [comment]
I agree, just storing the `Type` in `Name` will surely be enough.
I don't understand what you meant by:
>Of course because Type by itself doesn't define much we will have to support only a limited number of implementation / subclasses (probably only java.lang.Class and java.lang.reflect.ParameterizedType) but I think it's closer to core java API and more extensible.

Can you precise a little bit more ?

For the `Types` didn't remember its existence, although I tend to prefer something like `TypeReference`, it seems more fluent to me:
```java
Name.of(new TypeReference<Builder<A>>() {}, "builderA");
```
than
```java
Name.of(Types.newParameterizedType(Builder.class, A.class), "builderA");
``` 
But both signature could be added.

Currently I'm trying to figure out how I will be able to know that `Builder` is a "sub-type" of `Builder<A>`.
As I think it might be nice if, for example, those components are provided:
```java
@Provides @Named("builderA")
public Builder<A> builderA() {...}

@Provides @Named("anotherABuilder")
public Builder<A> anotherBuilder() {...}

@Provides @Named("builderB")
public Builder<B> builderB() {...}
```
To be able to make those queries:
- `factory.getComponents(Builder.class)` to get all builders, builderA, anotherABuilder and builderB
- `factory.getComponents(new TypeReference<Builder<A>() {})` to get all builders of type `Builder<A>`, builderA and anotherABuilder

And going a little bit further if we have class `A` defined as a super class of `B`:
```java
public class B extends A {
...
}
```
The second request `factory.getComponents(new TypeReference<Builder<A>() {})` should return the three builders, builderA, anotherABuilder and builderB.

**Comment by a-peyrard** [comment]
Forgot the "going further" part, I wasn't thinking clearly, cause, Java will not allow to put a `Builder<B>` in a reference of type `Builder<A>`.

**Comment by xhanin** [comment]
About my comment on `Type` being too generic, I'm just saying that if we accept a `Name.of(type1, "test")` then we won't be able to query it by class neither by parameterized type, and Factory implementations may / will be limited to some `Type` only.

For instance if you take current `SingleNameFactoryMachine` implementation, you have:
```
public boolean canBuild(Name<?> name) {
        return (this.name.equals(name)
                || (   this.name.getName().equals(name.getName())
                    && name.getClazz().isAssignableFrom(this.name.getClazz())));
}
```

This could be generalized for the `ParameterizedType`, but it won't handle any `Type` implementation.

Does this makes sense?

**Comment by xhanin** [comment]
About TypeReference readability I agree, that's why I said it was good to support it as a way to provide the `ParameterizedType`. But maybe in generated code we could use `Types` method which should be more efficient (it doesn't create a different class for each type reference).

For the "subtype" problem, I thnik we should have only limited support: if you ask for raw type you get all the parameterized types, if you ask for a parameterized type you get only components with that exact parameterized type, and that's it. For the rest it could come later but it's not easy to do correctly and people (including me) tend to be confused by generics inheritance.

**Comment by a-peyrard** [comment]
Ok, if I understand this right ;) this is the same concern as my last comment about getting to know if a type is a "sub-type" of another type (in reference to your second to last comment).
We need something like the method `isAssignableFrom` for types. Expecting those results:
```java
Type string = String.class;
Type charSequence = CharSequence.class;
isAssignableFrom(charSequence, string) == true

Type list = List.class
Type listOfString = new TypeReference<List<String>>() {}.getType();
Type arrayListOfString = new TypeReference<ArrayList<String>>() {}.getType();
isAssignableFrom(list, listOfString) == true
isAssignableFrom(list, arrayListOfString) == true
isAssignableFrom(listOfString, arrayListOfString) == true
```

And I agree for the `Types` in generated code, it might reduce the overhead that this feature will introduce.

**Comment by xhanin** [comment]
Agreed, fine for me. I wonder where the `isAssignableFrom` method should go? Maybe in the `Types` class?

**Comment by a-peyrard** [comment]
Yes I agree to put this method in `Types`, just need to find how to implement this ;)

Maybe just using `Class.isAssignableFrom` on rawtypes extracted from types should be enough... I would try this this evening, if I manage to find some free time.

**Comment by fcamblor** [comment]
Could we imagine to split this issue into 2 issues ?
I'm particularly interested by the `TypeReference` type (and maybe utility methods as well :-)) for #146 

I just ran into a compilation error while generating this code in routers :
```
        cacheParameterMapper(paramMapperRegistry, new EndpointParameter("POST", "/@/i18n/messages/{locale}", "entries"),
                java.util.Map<java.lang.String,java.lang.String>.class, EndpointParameterKind.BODY);
```
Cannot call `java.util.Map<java.lang.String,java.lang.String>.class` on parameterized types (thus, It would be nicer to use `TypeReference` here because I'll need to know parameterized types for params conversions)

**Comment by a-peyrard** [comment]
Ok, no problem

**Comment by a-peyrard** [comment]
Just a quick update, I did implement `TypeReference`: https://github.com/a-peyrard/restx/commit/0b3d9cce3f3fd906b414d27afe44616edd997f17
Add a `Types.getRawType` utility method: https://github.com/a-peyrard/restx/commit/5cec30e2730e51316ec5ba06bef2237f06aca54a
And I'm currently struggling with the implementation of `Types.isAssignableFrom`, I already wrote test cases (https://github.com/a-peyrard/restx/commit/5cec30e2730e51316ec5ba06bef2237f06aca54a), but maybe I went to far :) Especially with: 
```java
public static class GenericTypedImpl<T> implements GenericInterface<String> {}
```

**Comment by xhanin** [comment]
I think there is a missing commit, when you talk about `Types.isAssignableFrom` you reference the same commit as `Types.getRawType`.

And I think we don't need to get too far yet, it's better to have something limited in the first instance to see how useful it is, and then later come back to it to improve it.

**Comment by a-peyrard** [comment]
Indeed, the last one was not supposed to be a commit but a gist showing the tests I prepared for the `Types.isAssignableFrom` method. I just updated the test, so here it is: https://gist.github.com/a-peyrard/a07916d2ecd20f6af220

I will try to finish my implementation, if I did not manage to do it quickly I will reduce the perimeter, but I think I almost know where I'm going ;)

**Comment by a-peyrard** [comment]
So I just finished my very basic implementation of `Types.isAssignableFrom` ;) 
https://github.com/a-peyrard/restx/commit/298ee2332b480d892a472169cd62fa376835c27f

I'm a little bit concern of the overhead that will introduce the use of this method comparing to `Class.isAssignableFrom`. There is a lot of reflection done down here, and it might be way more time consuming than the "native" `Class.isAssignableFrom`. I will try to benchmark this, in order to have a better idea of this overhead.

WDTY ?

**Comment by xhanin** [comment]
That sounds good to me, good job! About performance, what is important is to preserve performance for the simple case of raw `Class` types that was the only one supported before. Checking the code I think it should be fast, but you're right a benchmark is interesting to have a clear idea.

**Comment by a-peyrard** [comment]
So I just implement a small benchmark: https://gist.github.com/a-peyrard/8160a7561f2fdb445337
Results are:
```
Benchmark                                                             Mode  Samples         Score        Error  Units
i.r.b.IsAssignableFromBenchmark.blank                                thrpt       15  41521288.041 ± 918186.599  ops/s
i.r.b.IsAssignableFromBenchmark.classIsAssignableFrom                thrpt       15  21441421.514 ± 213645.042  ops/s
i.r.b.IsAssignableFromBenchmark.typesIsAssignableFrom                thrpt       15  19478316.755 ± 133462.342  ops/s
i.r.b.IsAssignableFromBenchmark.typesWithGenericsIsAssignableFrom    thrpt       15   1356290.512 ±  45415.123  ops/s
```
The overhead for raw types is about 9%. Of course when not using raw types, it takes about 16x more time. 
For me it seems ok, WDYT ?

Just a quick remark about TypeReference implementation, if it's used without generics like this:
```java
new TypeReference() {};
```
An exception will be thrown.

In the jackson's `TypeReference` implementation, they are using a hack by implementing `Comparable`, in order to get a compilation error, but I guess it was working on jdk prior to 7, as it had no effect when I tried it.

**Comment by a-peyrard** [comment]
@fcamblor I just created a new PR with the current work on types: #166

If everyone is ok with this I will merge it.

**Comment by xhanin** [comment]
Agreed on the performance, considering how frequently this is used that should not be noticeable in most apps.

**Comment by a-peyrard** [comment]
Some update about this feature... It's almost done, here is the current branch https://github.com/a-peyrard/restx/tree/generics-in-dependency-injection (I just need to review the code in order to check if there is some missing things).
I hope I will find some time to submit the PR by the end of the week.

**Comment by xhanin** [comment]
That sounds great!

Pretty overwhelmed at work at the time being, so I'm not sure when I'll be
able to review that work, hopefully this week-end.

**Comment by a-peyrard** [comment]
Just some few remarks about the feature:
- If you did checkout the branch last week, it has been rebased since, cause some import was missing in mustaches templates (don't really know how I missed that though), consider a reset of your local branch.
- Some factory methods have been added to create `Name` with types. All methods have the same order of parameters, the type/class/typereference and the string, but one of them has the string at the beginning to be able to use varargs, it's not very elegant, but don't know how to do it by other means.
https://github.com/a-peyrard/restx/blob/generics-in-dependency-injection/restx-factory/src/main/java/restx/factory/Name.java#L40
- When using some Name factory methods, the varargs one, or the one with the type, Name has to be casted, because java can not analyse the produced parameterized type (Or it might also be called with the type specified in diamonds): 
````java
Name<Generic<String>> name = (Name<Generic<String>>) Name.of("foo", Generic.class, String.class);
Name<Generic<String>> name = Name.<Generic<String>>of("foo", Generic.class, String.class);
```` 
So in order to create name with generic type, the recommended method will be:
````java
Name<Generic<String>> name = Name.of(new TypeReference<Generic<String>>() {}, "foo");
```` 
- I didn't check components Customizer, but I might guess there is some things to adapt here, in order to be able to customize parameterized components.
- Same thing go for the deactivation keys
- Last but not least, I used a custom snapshot of restx, this week at work, shipping this new feature, and it was working fine. And it brings some very nice new design capabilities for components.

That's all for now, I'm pretty sure I forgot a lot of things that I wanted to say (unfortunately I did not write them all during the development). As soon as they will pop out of my brain, I will update this thread.

Have a nice we.

**Comment by a-peyrard** [comment]
In my last message I was talking about limitation in java inference, I wrote my message a little bit too quickly because the examples where not rights.
I was also talking about deactivation key, they have been modified to use types instead of classes, and for customizer there was nothing to change.

So I will prepare the PR for this new feature, and I also wrote some explanation about java inference.

There is things that unfortunately java miss, it did not manage to infer:

**Chained calls:** (this functionnality is still not fixed in jse8, and as far as I saw, there is nothing planed for jse9)
So unlike I was saying in my last message,
```java
Name<Generic<String>> name = Name.of("foo", Generic.class, String.class);
```
did compile, and 
```java
Factory.Query<TestGenericInterface<Double>> testGenericComponent = 
  Factory.Query.byType(Types.newParameterizedType(TestGenericInterface.class, Double.class));
``` 
is compiling too. But
```java
Factory.Query<TestGenericInterface<Double>> testGenericComponent = 
  Factory.Query.byType(Types.newParameterizedType(TestGenericInterface.class, Double.class)).mandatory();
``` 
is not compiling. Because of the chained call to mandatory.
So in order to compile, the diamonds operator must be used (or a cast) like this:
```java
Factory.Query<TestGenericInterface<Double>> testGenericComponent = 
  Factory.Query.<TestGenericInterface<Double>>byType(Types.newParameterizedType(TestGenericInterface.class, Double.class)).mandatory();
``` 

To continue on the same game, of "compile vs not compile".
```java
Query<Generic<String>> query1 = factory.queryByType(Generic.class, String.class);
Query<Generic<String>> query2 = factory.queryByType(new TypeReference<Generic<String>>() {});
```
Both lines are compiling.

But when trying to get the component by a fluent call to `findOneAsComponent`
```java
Optional<Generic<String>> c1 = factory.queryByType(Generic.class, String.class).findOneAsComponent();
Optional<Generic<String>> c2 = factory.queryByType(new TypeReference<Generic<String>>() {}).findOneAsComponent();
```
Only the second notation is compiling. For the first line, either an explicit cast or the diamonds operators must be used.

So when dealing with query like this, using the public API, I think it's easier to use the signature with the `TypeReference` (even if it has the overhead of a new anonymous type been created)

(some reading about this: http://openjdk.java.net/jeps/101 http://blog.jooq.org/2013/11/25/a-lesser-known-java-8-feature-generalized-target-type-inference/)

Another thing that I discovered, is the use of inference in iterator for loop, as I guess behind the scene there is a chained call to `iterator()`:
Something like this:
```java
Set<Name<T>> names = factory.collectAllBuildableNames(componentClass);
for (Name<T> tName : names) { 
```
can not be inlined without specifying the type with diamonds operators:
```java
for (Name<T> tName : factory.<T>collectAllBuildableNames(componentClass)) {
```

**Comment by a-peyrard** [comment]
Some other things that java does not infer:
```java
Generic<String> foo = 
    factory.getComponent(Name.of("foo", Generic.class, String.class));
```
So it has to be written like this:
```java
Generic<String> foo = 
    factory.getComponent(Name.<Generic<String>>of("foo", Generic.class, String.class));
```
(I did not try but I guess this problem is solved in jse8)

Using `TypeReference` the problem no longer exist:
```java
Generic<String> foo = 
    factory.getComponent(Name.of(new TypeReference<Generic<String>>() {}, "foo"));
```
